ash-window/src/lib.rs

```diff
@@ -20,77 +23,59 @@ use ash::extensions::ext; // portability extensions
 pub unsafe fn create_surface(
     entry: &Entry,
     instance: &Instance,
+    // TODO: Cannot write `&dyn HasRawDisplayHandle + HasRawWindowHandle :/
+    display_handle: &dyn HasRawDisplayHandle,
     window_handle: &dyn HasRawWindowHandle,
```

Comment:
@kchibisov what was the idea here?

Reply:
I think you can combine them, or what is the issue here?

Reply:
> what is the issue here?

```rust
error[E0225]: only auto traits can be used as additional traits in a trait object
  --> ash-window/src/lib.rs:26:41
   |
26 |     handle: &(dyn HasRawDisplayHandle + HasRawWindowHandle),
   |                   -------------------   ^^^^^^^^^^^^^^^^^^ additional non-auto trait
   |                   |
   |                   first non-auto trait
   |
   = help: consider creating a new trait with all of these as supertraits and using that trait here instead: `trait NewTrait: HasRawDisplayHandle + HasRawWindowHandle {}`
   = note: auto-traits like `Send` and `Sync` are traits that have special properties; for more information on them, visit <https://doc.rust-lang.org/reference/special-types-and-traits.html#auto-traits>
```

Reply:
This works:

```rust
pub trait RawHandles: HasRawDisplayHandle + HasRawWindowHandle {
    fn get(&self) -> (RawDisplayHandle, RawWindowHandle);
}

impl<T: HasRawDisplayHandle + HasRawWindowHandle> RawHandles for T {
    fn get(&self) -> (RawDisplayHandle, RawWindowHandle) {
        (self.raw_display_handle(), self.raw_window_handle())
    }
}

pub unsafe fn create_surface(
    ...,
    handle: &dyn RawHandles,
    ...,
) {
    match handle.get() {
        (RawDisplayHandle..., RawWindowHandle...) =>
    }
}
```

Reply:
I think you can `Box` them?

Reply:
@kchibisov Same error... But even if it did that defeats the purpose of stuffing a cheap borrow of a `Window` in there as the user now has to _own_ (or clone?) it and temporarily move it into a heap allocation unnecessarily.

I'd then rather change the API to accept `RawDisplayHandle` and `RawWindowHandle` by-value.

Reply:
switching to static dispatching would be also an option (which wgpu will do https://github.com/gfx-rs/wgpu/blob/ab4df8bcbfd937a9fb527f797590d559c13076bb/wgpu-core/src/instance.rs#L427)

Reply:
@msiglreith Cool, there's probably no harm in doing static dispatching in `ash-window` either?

Reply:
Hmm, I'm currently using `dyn` dispatch on my side within trait functions (i.e. `trait Renderer { fn create_swapchain(&dyn HasRawWindowHandle, ...) }`), that'd be annoying.

Reply:
@msiglreith Fwiw @i509VCB recently (3 days ago) changed this to use the owned/resolved `RawDisplayHandle`/`RawWindowHandle`): https://github.com/gfx-rs/wgpu/commit/9e3efd7019518d372516820c77d4456ae2882d99

That way the caller can simply call `.raw_display_handle()`/`.raw_window_handle()`, without having to worry that - for example - the `Has*` traits are not implemented for those types.

Reply:
Implemented it that way now, as originally proposed in https://github.com/ash-rs/ash/pull/645#discussion_r933060153 too, so we're back at static dispatch again at the detriment of requiring the caller to now add/use the `raw-window-handle` crate.

Fortunately these `Raw*Handle` types are `Copy`, so it's perfectly fine to move them around instead of borrowing too.